nucliadb/nucliadb/purge/orphan_shards.py

```diff
@@ -0,0 +1,198 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+from dataclasses import dataclass
+
+import pkg_resources
+from grpc.aio import AioRpcError  # type: ignore
+
+from nucliadb.common.cluster import manager
+from nucliadb.common.cluster.exceptions import ShardsNotFound
+from nucliadb.common.cluster.manager import KBShardManager
+from nucliadb.common.cluster.utils import setup_cluster, teardown_cluster
+from nucliadb.common.maindb.driver import Driver
+from nucliadb.common.maindb.utils import setup_driver, teardown_driver
+from nucliadb.ingest import logger
+from nucliadb.ingest.orm.knowledgebox import KnowledgeBox
+from nucliadb_telemetry import errors
+from nucliadb_telemetry.logs import setup_logging
+from nucliadb_utils.settings import running_settings
+
+
+@dataclass
+class ShardLocation:
+    kbid: str
+    node_id: str
+
+
+async def detect_orphan_shards(driver: Driver):
+    """Detect orphan shards in the system. An orphan shard is one indexed but
+    not referenced for any stored KB.
+
+    """
+    # To avoid detecting a new shard as orphan, query the index first and maindb
+    # afterwards
+    indexed_shards = await _get_indexed_shards()
+    stored_shards = await _get_stored_shards(driver)
+
+    # Log an error in case we found a shard stored but not indexed, this should
+    # never happen as shards are created in the index node and then stored in
+    # maindb
+    not_indexed_shards = stored_shards.keys() - indexed_shards.keys()
+    for shard_id in not_indexed_shards:
+        location = stored_shards[shard_id]
+        logger.error(
+            "Found a shard on maindb not indexed in the index nodes",
+            extra={
+                "shard_id": shard_id,
+                "kbid": location.kbid,
+                "node_id": location.node_id,
+            },
+        )
+
+    orphan_shard_ids = indexed_shards.keys() - stored_shards.keys()
+    orphan_shards: dict[str, ShardLocation] = {
+        shard_id: indexed_shards[shard_id] for shard_id in orphan_shard_ids
+    }
+    return orphan_shards
+
+
+async def _get_indexed_shards() -> dict[str, ShardLocation]:
+    indexed_shards: dict[str, ShardLocation] = {}
+    available_nodes = manager.get_index_nodes()
+    for node in available_nodes:
+        node_shards = await node.list_shards()
+        for shard_id in node_shards:
+            try:
+                shard_pb = await node.get_shard(shard_id)
```

Comment:
I'm worried about this, as this will bring the shard into the node memory cache right? is it needed?

Maybe we should only do this check for the shards that we detected as orphaned/ missing, which hopefully it is a smaller subset?

Reply:
It's the only supported operation right now that returns the `kbid`. As it's not something you'll run often, I thought it wouldn't be a problem. But I'll change it to do it only for orphan shards.